### ISSUE TITLE
Add support for IPv4 for 0-OS workloads

### ISSUE THREAD

**Issue by robvanmieghem** [threefoldtech/home#925]
At the moment there is no way to assign a public IPv4 address to any 0-OS workloads. While this is mitigated with the presence of the web gateway, there are situation where being able to publicly expose services directly is preferable.

This story aims to address this by implementing the following:
- [x] Specification for the internal zos plumbing [@delandtj] https://github.com/threefoldtech/zos/issues/1051
- [x] Create a new 0-OS primitive that represent an IPv4 public address.
  - [x] Update the definition of k8s to support linking a IP workload to it
  - [x] Update 0-OS to support the new IP primitive
  - [x] Allow the farmer to define pools of IP usable for renting.
- [x] In the 1st phase, only allow to link an IP address to a K8S VM. (eventually this could also be attached to container, but on a later phase)

**Comment by muhamadazmy** [comment]
## Specifications
### General Idea
- Farmers must maintain a pool of public IPv4 that are attached to a certain farm
  - Do nodes need to know about possible IPs ? (in case a farm is spread over multiple geological locations)
- IP reservation
  - IP reservation is not related to a certain K8S, or container reservation.
  - IP reservation shouldn't be attached to a node-id instead should be attached to a farmer id. Deployment of the IP reservation follow the k8s (or container) node id. (this is totally not supported atm)
  - IP reservation expiration must be greater than or equal to the expiration of the K8S reservation
  - When an IP reservation expires, does it go back to the pool immediately? 
- Attach an IP to k8s reservation
  - Validate that it's not already used (over the explorer?)
- Node must validate that is's a valid Ip address from the farm pool
- Setup plumbing 

## Proposal (iteration no. 1)
- Farmer must have a stand alone pool manager that does the following (can be part of the farmer bot)
 - Process farmer ip reservation
 - Allocate IPs in a race free manner and answer the reservation with the valid IP address
 - track reservations expirations 
 - Once an IP reservation has expired, it must be either:
   - Immediately go back to the pool
   - Or put back to a cold storage, IPs here are not ready for immediate reuse expect if requested by the same previous owner, until it hit a configurable expiration time, then go back to the free pool.

**Comment by robvanmieghem** [comment]
@muhamadazmy  Looks good but in the first iteration I would drop " Or put back to a cold storage, IPs here are not ready for immediate reuse expect if requested by the same previous owner, until it hit a configurable expiration time, then go back to the free pool."

**Comment by zaibon** [comment]
> Farmers must maintain a pool of public IPv4 that are attached to a certain farm

Agreed. A farm today represent a bunch of node located near each other usually using the same network infrastructure. Linking the IP pool to a farm would allow any node of the farm to use any of the IP from the pool.

> Farmer must have a stand alone pool manager that does the following (can be part of the farmer bot)

Yeah this needs to be though out carefully. If we force farmer to have a 3bot to handle this kind of things, we better make sure each farmer is able to have one and it is stable enough. But I agree that it would be best to make it live in the farmer then on the explorer. 

> Do nodes need to know about possible IPs ?

They should not. A node will received an IP reservation like any other reservation which must contain all the information required to configure the IP address on the required NIC.


> IP reservation is not related to a certain K8S, or container reservation.

Agreed, the IP primitive should be a primitive by itself that can then later by link to a K8S VM or container, the same way volume works today.

> IP reservation shouldn't be attached to a node-id instead should be attached to a farmer id. Deployment of the IP reservation follow the k8s (or container) node id. (this is totally not supported atm)

I don't see how that would work. A reservation is always linked to a node. Why would we need to change this ?


> IP reservation expiration must be greater than or equal to the expiration of the K8S reservation

While expiration is not a thing anymore. User should make sure the pool attached to its IP should be valid for as long as its attched VM/Container. I do not think this should be managed by 0-OS though. This sounds like a more general management that needs to be done by the 3Bot/User.

> When an IP reservation expires, does it go back to the pool immediately?

Yes. As soon as the node mark the IP reservation to be `DELETED`. Then it should also somehow notify the farm that the IP is available again.

**Comment by muhamadazmy** [comment]
After the mgmt call, we agreed to introduce a farmer bot to manage both pricing and IP config, the bot will be reachable over libp2p (to be speced)

**Comment by muhamadazmy** [comment]
## Proposal (iteration 2)
- User creates an IPv4 reservation on a certain node (normally like any other reservation)
- The node starts processing this reservation normally like any other reservation
- The handler for this reservation does the following:
  - Contact the farmer bot (reachable by knowing the farmer id, hence the libp2p address of the bot
  - The bot replies with a free IP address from a configurable pool.
  - The node update the reservation object with the results (success or failure)
  - Node attach the IP address to user NR (network resource) [TO BE SPEC-ED]
- Once reservation is successful, the user now can associate this IP to his k8s reservation 
  - approach 1:
    - k8s node must be created on the same node as the IP and then traffic is forwarded internally to the vm internal IP address
  - approach 2:
    - k8s node can be created on ANY node that is part of the user network, and then traffic can be routed over wireguard 
> Again: to be speced which of these approaches are doable. Also if forwarding is possible, then public IP can also be easily used with containers.

**Comment by delandtj** [comment]
I'm still wondering if putting the IP in an NR is a good idea. 
pros: 
  - security-wise it's better, as the vm still doesn't have a public ip, nor can it spoof one
  - no need to adapt k3s image to be able to handle multi-homed ip 
  - no complex routing setup in ZOS, a part from movint the deft gw to the public space
cons:
  - k8s has capability to move IP to another node in the same segment in case a node would fail
  - k8s it's load balancer features are specifically designed to be efficient

**Comment by LeeSmet** [comment]
After some talk with @delandtj and seeing what is needed for the reservation of an IP, it seems we need 2 things in zos: the IP itself, and a default gateway. The default gateway is known on the node, since it can be extracted from whatever namespace it currently resides in (ndmz, public, ...?). This means we just need to inform the node about the IP. 

As such, I think it is best to have a list of public IP's available in the farm on the explorer, along with a bool for every one whether it is taken or not. This means we won't need a farmer bot (specifically for this part), and the whole reservation flow is in line with the other primitives we already have. A customer simply makes a reservation for an IP in a farm. Once the IP is reserved, we can allow it to be added to other reservations (just k8s for now) by adding the reservation ID in the new reservation (similar to how volumes work now).

 All the explorer has to do in this case is verify that the IP is not yet taken, which is very acceptable. This is also a setup which is perfectly portable to a blockchain setup in the future.

**Comment by muhamadazmy** [comment]
@LeeSmet yes, I was actually thinking exactly the same scenario to avoid the farmer bot. I will start creating the tasks as per this setup.

**Comment by muhamadazmy** [comment]
@LeeSmet Regarding the k8s deployment part. I was thinking if as discussed with @delandtj if we can define a mapping from public ip to private ip in the reservation, then we don't need to change how the k8s node (or a container). The ip traffic routing can live independent from the destination node.

Once the ip reservation is deployed on the node the node will build the proper routing, and forwarding rules to route the traffice from that public ip to a user private ip on the wireguard (through the NR)

then the k8s (or container) can live anywhere since we select its private ip on deployment

**Comment by delandtj** [comment]
OK, 
things to do :

## 0-OS
  - [ ] 0-OS prepare virtio-fs files for Rancher to get `/etc/resolv.conf , /etc/hostname , connman.conf` into a prepared  Rancher instance.
    - [ ] prepare Cloud-Hypervisor CI and flist
    - [ ] adapt 0-OS to use CH instead of Firecracker
    - [ ] adapt Rancher image to temp mount virtio-fs and copy files in place before pivot-root

   - [ ] get adapted workload containing Public IPv4 Addr, default gateway, and dns server config

**Comment by DylanVerstraete** [comment]
Estimate for Explorer / Client side: 2-3 days to finish and test

Things to do still:

- [ ] review data model
- [ ] review API routes and business logic
- [ ] add new RSU type for network (Ipaddress) type
- [ ] calculate how it would cost for an ip for x time
- [ ] client tools (jsdk) already in progress

Optional

- [ ] UI in farmer dashboard so a farmer can add ipaddresses to it's farm
- [ ] ..

**Comment by LeeSmet** [comment]
I manually updated `vmd`, `networkd` and `provisiond` of 1 devnet node based on the `vm_public_ip` branch on zos. This should have all required changes to enable public ipv4 inside a vm, as well as (private) ipv6. However since js-sdk refuses to launch on my machine, I can't test the changes at this time to merge the branch and updated all development nodes

**Comment by xmonader** [comment]
@LeeSmet please open an issue on js-sdk repo with a detailed description of the problem.

**Comment by LeeSmet** [comment]
Original problem: threefoldtech/js-sdk/issues/1713, a workaround is described in the issue.

After that, I'm now stuck on threefoldtech/js-sdk/issues/1714

**Comment by LeeSmet** [comment]
After a lot of messing around I was able to deploy a k3s vm with a public IP address:

```
🔋47% 🕙[ 16:49:37 ] ❯ ssh rancher@185.69.167.121
The authenticity of host '185.69.167.121 (185.69.167.121)' can't be established.
ECDSA key fingerprint is SHA256:FWnYbLo445VjMeuHjTtF6G+FGzP9x9dYK5gfNjY9z4M.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '185.69.167.121' (ECDSA) to the list of known hosts.
Welcome to k3OS!

Refer to https://github.com/rancher/k3os for README and issues

By default mode of k3OS is to run a single node cluster. Use "kubectl"
to access it.  The node token in /var/lib/rancher/k3s/server/node-token
can be used to join agents to this server.

k3os-19211 [~]$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether f6:e2:0d:c4:67:a1 brd ff:ff:ff:ff:ff:ff
    inet 10.20.1.50/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fd41:6662:6f48:1::32/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::f4e2:dff:fec4:67a1/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 2a:ac:53:59:4d:36 brd ff:ff:ff:ff:ff:ff
    inet 185.69.167.121/24 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 2a02:1802:5e:0:28ac:53ff:fe59:4d36/64 scope global dynamic mngtmpaddr 
       valid_lft 43143sec preferred_lft 43143sec
    inet6 fe80::28ac:53ff:fe59:4d36/64 scope link 
       valid_lft forever preferred_lft forever
4: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
5: gre0@NONE: <NOARP> mtu 1476 qdisc noop state DOWN group default qlen 1000
    link/gre 0.0.0.0 brd 0.0.0.0
6: gretap0@NONE: <BROADCAST,MULTICAST> mtu 1462 qdisc noop state DOWN group default qlen 1000
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
7: erspan0@NONE: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN group default qlen 1000
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
8: ip_vti0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
9: ip6_vti0@NONE: <NOARP> mtu 1428 qdisc noop state DOWN group default qlen 1000
    link/tunnel6 :: brd :: permaddr dece:a7dd:94b4::
10: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/sit 0.0.0.0 brd 0.0.0.0
11: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN group default qlen 1000
    link/tunnel6 :: brd :: permaddr fe08:903d:1469::
12: ip6gre0@NONE: <NOARP> mtu 1448 qdisc noop state DOWN group default qlen 1000
    link/gre6 :: brd :: permaddr aa4e:d9c3:6ac7::
13: flannel.1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UNKNOWN group default 
    link/ether 32:90:e8:e9:f6:32 brd ff:ff:ff:ff:ff:ff
    inet 10.42.0.0/32 scope global flannel.1
       valid_lft forever preferred_lft forever
    inet6 fe80::3090:e8ff:fee9:f632/64 scope link 
       valid_lft forever preferred_lft forever
```

As of right now, public ip reservations on devnet are possible, but only VM's on node `2anfZwrzskXiUHPLTqH1veJAia8G6rW2eFLy5YFMa1MP` will be able to use them. A lot of stuff still needs to be fixed/improved before this can be merged.

Work is also done on ipv6 in the vm, though this is currently not stable yet and probably unusable

**Comment by xmonader** [comment]
@muhamadazmy can we please update the story body to reflect the correct checklist

**Comment by m-motawea** [comment]
is there a limitation on nodes that can be used to deploy ip addresses? like a node with public ipv4 endpoint?

**Comment by LeeSmet** [comment]
IPv4 support is available (still on the one node, but the branch should ideally be merged today and all of devnet upgraded).

Right now we are preparing to add yggdrasil support as well.
Simultaniously, the network setup script in the vm is being improved.

Once the PR is merged, we will work on having yggdrasil IP's inside of the container

**Comment by LeeSmet** [comment]
> is there a limitation on nodes that can be used to deploy ip addresses? like a node with public ipv4 endpoint?

Not all nodes will support this. For example, hidden nodes will not. The easiest way to identify nodes that support this will probably be to iterate over the interfaces of the node and look for the existence of a bridge named "br-pub".

Also, the farm of the node will obviously need to have public IP's available

**Comment by LeeSmet** [comment]
devnet has been upgraded with the new networking code. The upper half (10.6.0.{11-14}) has been rebooted, public IP is now available on all these 4 nodes. the lower half will be rebooted next week, after we validated that the new networking code is backward compatible. 

The pr also has some stability fixes for issues discovered on the previously patched node

**Comment by xmonader** [comment]
@LeeSmet https://explorer.devnet.grid.tf/ isn't reachable

**Comment by m-motawea** [comment]
> Not all nodes will support this. For example, hidden nodes will not. The easiest way to identify nodes that support this will probably be to iterate over the interfaces of the node and look for the existence of a bridge named "br-pub".

this seems to be ok in case there is only one public subnet. but if there are multiple public subnets, how to know if the node has access to the gateway of the address I'm trying to reserve?

**Comment by muhamadazmy** [comment]
> > Not all nodes will support this. For example, hidden nodes will not. The easiest way to identify nodes that support this will probably be to iterate over the interfaces of the node and look for the existence of a bridge named "br-pub".
> 
> this seems to be ok in case there is only one public subnet. but if there are multiple public subnets, how to know if the node has access to the gateway of the address I'm trying to reserve?

I am not entirely sure but this should be handled by the farm switches, for you as a user it should not matter, @LeeSmet please confirm ?
